// umi/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  Link: true,
  NavLink: true,
  Redirect: true,
  dynamic: true,
  router: true,
  withRouter: true,
  Route: true
};
Object.defineProperty(exports, "Link", {
  enumerable: true,
  get: function get() {
    return _link.default;
  }
});
Object.defineProperty(exports, "NavLink", {
  enumerable: true,
  get: function get() {
    return _navlink.default;
  }
});
Object.defineProperty(exports, "Redirect", {
  enumerable: true,
  get: function get() {
    return _redirect.default;
  }
});
Object.defineProperty(exports, "dynamic", {
  enumerable: true,
  get: function get() {
    return _dynamic.default;
  }
});
Object.defineProperty(exports, "router", {
  enumerable: true,
  get: function get() {
    return _router.default;
  }
});
Object.defineProperty(exports, "withRouter", {
  enumerable: true,
  get: function get() {
    return _withRouter.default;
  }
});
Object.defineProperty(exports, "Route", {
  enumerable: true,
  get: function get() {
    return _Route.default;
  }
});

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function _react() {
    return data;
  };

  return data;
}

var _link = _interopRequireDefault(require("./link"));

var _navlink = _interopRequireDefault(require("./navlink"));

var _redirect = _interopRequireDefault(require("./redirect"));

var _dynamic = _interopRequireDefault(require("./dynamic"));

var _router = _interopRequireDefault(require("./router"));

var _withRouter = _interopRequireDefault(require("./withRouter"));

var _Route = _interopRequireDefault(require("./Route"));

var _umiExports = require("@tmp/umiExports");

Object.keys(_umiExports).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _umiExports[key];
    }
  });
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
