// umi/lib/runtimePlugin.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.init = init;
exports.use = use;
exports.getItem = getItem;
exports.compose = compose;
exports.apply = apply;
exports.applyForEach = applyForEach;
exports.mergeConfig = mergeConfig;
exports.mergeConfigAsync = mergeConfigAsync;

var _assert = _interopRequireDefault(require("assert"));

var _isPlainObject = _interopRequireDefault(require("lodash/isPlainObject"));

var _utils = require("./utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

var plugins = null;
var validKeys = [];

function init() {
  var opts = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  plugins = [];
  validKeys = opts.validKeys || [];
}

function use(plugin) {
  Object.keys(plugin).forEach(function (key) {
    // TODO: remove default
    // default 是为了兼容内部框架内置的一个 babel 插件问题
    (0, _assert.default)(validKeys.concat('default').indexOf(key) > -1, "Invalid key ".concat(key, " from plugin"));
  });
  plugins.push(plugin);
}

function getItem(key) {
  (0, _assert.default)(validKeys.indexOf(key) > -1, "Invalid key ".concat(key));
  return plugins.filter(function (plugin) {
    return key in plugin;
  }).map(function (plugin) {
    return plugin[key];
  });
}

function _compose() {
  for (var _len = arguments.length, funcs = new Array(_len), _key = 0; _key < _len; _key++) {
    funcs[_key] = arguments[_key];
  }

  if (funcs.length === 1) {
    return funcs[0];
  }

  var last = funcs.pop();
  return funcs.reduce(function (a, b) {
    return function () {
      return b(a);
    };
  }, last);
}

function compose(item, _ref) {
  var initialValue = _ref.initialValue;
  if (typeof item === 'string') item = getItem(item);
  return function () {
    return _compose.apply(void 0, _toConsumableArray(item).concat([initialValue]))();
  };
}

function apply(item, _ref2) {
  var initialValue = _ref2.initialValue,
      args = _ref2.args;
  if (typeof item === 'string') item = getItem(item);
  (0, _assert.default)(Array.isArray(item), "item must be Array");
  return item.reduce(function (memo, fn) {
    (0, _assert.default)(typeof fn === 'function', "applied item must be function");
    return fn(memo, args);
  }, initialValue);
}

function applyForEach(item, _ref3) {
  var initialValue = _ref3.initialValue;
  if (typeof item === 'string') item = getItem(item);
  (0, _assert.default)(Array.isArray(item), "item must be Array");
  item.forEach(function (fn) {
    (0, _assert.default)(typeof fn === 'function', "applied item must be function");
    fn(initialValue);
  });
} // shadow merge


function mergeConfig(item) {
  if (typeof item === 'string') item = getItem(item);
  (0, _assert.default)(Array.isArray(item), "item must be Array");
  return item.reduce(function (memo, config) {
    (0, _assert.default)((0, _isPlainObject.default)(config), "Config is not plain object");
    return _objectSpread({}, memo, {}, config);
  }, {});
}

function mergeConfigAsync(item) {
  var mergedConfig, _iteratorNormalCompletion, _didIteratorError, _iteratorError, _iterator, _step, config;

  return regeneratorRuntime.async(function mergeConfigAsync$(_context) {
    while (1) {
      switch (_context.prev = _context.next) {
        case 0:
          if (typeof item === 'string') item = getItem(item);
          (0, _assert.default)(Array.isArray(item), "item must be Array");
          mergedConfig = {};
          _iteratorNormalCompletion = true;
          _didIteratorError = false;
          _iteratorError = undefined;
          _context.prev = 6;
          _iterator = item[Symbol.iterator]();

        case 8:
          if (_iteratorNormalCompletion = (_step = _iterator.next()).done) {
            _context.next = 19;
            break;
          }

          config = _step.value;

          if (!(0, _utils.isPromiseLike)(config)) {
            _context.next = 14;
            break;
          }

          _context.next = 13;
          return regeneratorRuntime.awrap(config);

        case 13:
          config = _context.sent;

        case 14:
          (0, _assert.default)((0, _isPlainObject.default)(config), "Config is not plain object");
          mergedConfig = _objectSpread({}, mergedConfig, {}, config);

        case 16:
          _iteratorNormalCompletion = true;
          _context.next = 8;
          break;

        case 19:
          _context.next = 25;
          break;

        case 21:
          _context.prev = 21;
          _context.t0 = _context["catch"](6);
          _didIteratorError = true;
          _iteratorError = _context.t0;

        case 25:
          _context.prev = 25;
          _context.prev = 26;

          if (!_iteratorNormalCompletion && _iterator.return != null) {
            _iterator.return();
          }

        case 28:
          _context.prev = 28;

          if (!_didIteratorError) {
            _context.next = 31;
            break;
          }

          throw _iteratorError;

        case 31:
          return _context.finish(28);

        case 32:
          return _context.finish(25);

        case 33:
          return _context.abrupt("return", mergedConfig);

        case 34:
        case "end":
          return _context.stop();
      }
    }
  }, null, null, [[6, 21, 25, 33], [26,, 28, 32]]);
}
