// umi-notify/index.js
const chalk = require('chalk');
const isAliEnv = require('is-ali-env');

function useBigfish(type, pkg) {
  if (process.env.MONITOR === 'none') return;
  isAliEnv().then(val => {
    if (val) {
      let stopMsg = pkg.name === 'roadhog'
        ? `${pkg.name} å·²äºŽ 2019.3.1 åœæ­¢ç»´æŠ¤`
        : `${pkg.name} å·²åœæ­¢ç»´æŠ¤`;
      console.log(`
---
ðŸ’¥ðŸ’¥ðŸ’¥ ${chalk.bold.red(`${stopMsg}ï¼Œä¸ºäº†æ›´å¥½å¯¹æ‚¨è¿›è¡ŒæœåŠ¡ï¼ŒæŽ¨èä½¿ç”¨ bigfishã€‚`)}

é’‰é’‰ç¾¤ï¼šBigfish é‡‘ç‰ŒæœåŠ¡ç¾¤ ï¼ˆç¾¤å·21761376ï¼‰
å®˜ç½‘ï¼šhttps://bigfish.alipay.com/
${pkg.name === 'roadhog' ? `å‡çº§æ–‡æ¡£ï¼šhttps://bigfish.alipay.com/doc/dgmrg0\n` : `å‡çº§æ–‡æ¡£ï¼šhttps://bigfish.alipay.com/doc/tnyo3y\n`}
---
`);
    }
  });
}

const messages = {
  buildComplete: {
    roadhog: [useBigfish],
    'atool-build': [useBigfish],
  },
  devStart: {
    dora: [useBigfish],
  },
  devComplete: {
    roadhog: [useBigfish],
    'atool-build': [useBigfish],
  },
};

function run(type, pkg, opts) {
  try {
    messages[type][pkg.name].forEach(message => {
      if (typeof message === 'string') {
        console.log(message);
      } else if (typeof message === 'function') {
        message(type, pkg, opts);
      }
    })
  } catch (e) {
  }
}

exports.onBuildStart = function (pkg, opts) {
  run('buildStart', pkg, opts);
}

exports.onBuildComplete = function (pkg, opts) {
  run('buildComplete', pkg, opts);
}

exports.onDevStart = function (pkg, opts) {
  run('devStart', pkg, opts);
}

exports.onDevComplete = function (pkg, opts) {
  run('devComplete', pkg, opts);
}
