// umi-ui-tasks/lib/ui/util/terminal.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getTerminalIns = void 0;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function _react() {
    return data;
  };

  return data;
}

var _enums = require("../../server/core/enums");

function _xtermAddonWebLinks() {
  const data = require("xterm-addon-web-links");

  _xtermAddonWebLinks = function _xtermAddonWebLinks() {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const _window = window,
      fit = _window.fit,
      Terminal = _window.Terminal;

function initTerminal() {
  Terminal.applyAddon(fit);
  const terminal = new Terminal({
    allowTransparency: true,
    theme: {
      background: '#15171C',
      foreground: '#ffffff73'
    },
    fontFamily: `operator mono,SFMono-Regular,Consolas,Liberation Mono,Menlo,monospace`,
    cursorBlink: false,
    cursorStyle: 'underline',
    disableStdin: true
  });
  terminal.loadAddon(new (_xtermAddonWebLinks().WebLinksAddon)());
  return terminal;
}

const TASKS = [_enums.TaskType.BUILD, _enums.TaskType.DEV, _enums.TaskType.LINT, _enums.TaskType.TEST, _enums.TaskType.INSTALL];
const TERMINAL_MAPS = {};
TASKS.forEach(taskType => {
  TERMINAL_MAPS[taskType] = initTerminal();
});

const getTerminalIns = taskType => TERMINAL_MAPS[taskType];

exports.getTerminalIns = getTerminalIns;
