// umi-plugin-react/lib/ui.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getPluginConfig = getPluginConfig;
exports.default = void 0;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function _react() {
    return data;
  };

  return data;
}

function _configuration() {
  const data = require("umi-plugin-ui/lib/plugins/configuration");

  _configuration = function _configuration() {
    return data;
  };

  return data;
}

var _config = require("./config");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { if (!(Symbol.iterator in Object(arr) || Object.prototype.toString.call(arr) === "[object Arguments]")) { return; } var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function getPluginConfig(plugins, name) {
  var _iteratorNormalCompletion = true;
  var _didIteratorError = false;
  var _iteratorError = undefined;

  try {
    for (var _iterator = plugins[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
      const plugin = _step.value;

      if (typeof plugin === 'string' && plugin === name) {
        return {};
      }

      if (Array.isArray(plugin) && plugin[0] === name) {
        if (typeof plugin[1] === 'object') {
          return plugin[1];
        } else {
          return {};
        }
      }
    }
  } catch (err) {
    _didIteratorError = true;
    _iteratorError = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion && _iterator.return != null) {
        _iterator.return();
      }
    } finally {
      if (_didIteratorError) {
        throw _iteratorError;
      }
    }
  }

  return null;
}

var _default = api => {
  if (process.env.UMI_UI === 'none') return;

  function getConfig(lang) {
    const userConfig = api.service.userConfig;

    const _userConfig$getConfig = userConfig.getConfig({
      force: true
    }),
          plugins = _userConfig$getConfig.plugins;

    const pluginConfig = getPluginConfig(plugins, 'umi-plugin-react');
    return (0, _configuration().formatConfigs)(_config.configs, {
      lang,
      groupMap: _config.groupMap
    }).map(p => {
      const _useConfigKey = (0, _configuration().useConfigKey)(pluginConfig, p.name),
            _useConfigKey2 = _slicedToArray(_useConfigKey, 2),
            haveKey = _useConfigKey2[0],
            value = _useConfigKey2[1];

      if (haveKey) {
        p.value = value;
      }

      return p;
    });
  }

  function validateConfig(config) {}

  if (api.addUIPlugin) {
    api.addUIPlugin(require.resolve('../ui/dist/index.umd'));
  }

  if (api.onUISocket) {
    api.onUISocket(({
      action,
      failure,
      success
    }) => {
      const type = action.type,
            payload = action.payload,
            lang = action.lang;

      switch (type) {
        case 'org.umi.umi-plugin-react.config.list':
          success({
            data: getConfig(lang)
          });
          break;

        case 'org.umi.umi-plugin-react.config.edit':
          let config = payload.key;

          if (typeof payload.key === 'string') {
            config = {
              [payload.key]: payload.value
            };
          }

          try {
            validateConfig(config);
            api.service.runCommand('config', {
              _: ['set', config],
              plugin: 'umi-plugin-react'
            });
            success({});
          } catch (e) {
            failure({
              message: e.message,
              errors: e.errors
            });
          }

          break;

        default:
          break;
      }
    });
  }
};

exports.default = _default;
