// umi-plugin-ui/lib/plugins/dashboard/ui/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _icons = require("@ant-design/icons");

var _antd = require("antd");

var _uiModule = _interopRequireDefault(require("./ui.module.less"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { if (!(Symbol.iterator in Object(arr) || Object.prototype.toString.call(arr) === "[object Arguments]")) { return; } var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var useState = _react.default.useState,
    useRef = _react.default.useRef,
    useEffect = _react.default.useEffect;

var DashboardUI = function DashboardUI(props) {
  // TODO react-grid-layout to custom layout
  var isClosed = window.localStorage.getItem('umi_ui_dashboard_welcome') || false;

  var _useState = useState(!!isClosed),
      _useState2 = _slicedToArray(_useState, 2),
      closed = _useState2[0],
      setClosed = _useState2[1];

  var api = props.api;
  var redirect = api.redirect,
      currentProject = api.currentProject,
      _ = api._,
      intl = api.intl,
      _api$getBasicUI = api.getBasicUI,
      getBasicUI = _api$getBasicUI === void 0 ? function () {
    return {};
  } : _api$getBasicUI;
  var FormattedMessage = intl.FormattedMessage;
  var actionCardCls = (0, _classnames.default)(_uiModule.default.card, _uiModule.default['card-action']);
  var welcomeCardCls = (0, _classnames.default)(_uiModule.default.card, _uiModule.default.welcome);
  var basicUI = getBasicUI();
  var actionCards = [{
    className: actionCardCls,
    title: _react.default.createElement("div", {
      className: _uiModule.default.main
    }, _react.default.createElement("div", {
      className: (0, _classnames.default)(_uiModule.default.icon, _uiModule.default.dev)
    }), _react.default.createElement("div", {
      className: _uiModule.default.info
    }, _react.default.createElement("h4", null, intl({
      id: 'org.umi.ui.dashboard.panel.dev.title'
    })), _react.default.createElement("p", null, intl({
      id: 'org.umi.ui.dashboard.panel.dev.desc'
    })))),
    body: _react.default.createElement("div", {
      onClick: function onClick() {
        return redirect('/tasks');
      }
    }, _react.default.createElement(FormattedMessage, {
      id: "org.umi.ui.dashboard.panel.goto.task"
    }))
  }, {
    className: actionCardCls,
    title: _react.default.createElement("div", {
      className: _uiModule.default.main
    }, _react.default.createElement("div", {
      className: (0, _classnames.default)(_uiModule.default.icon, _uiModule.default.build)
    }), _react.default.createElement("div", {
      className: _uiModule.default.info
    }, _react.default.createElement("h4", null, intl({
      id: 'org.umi.ui.dashboard.panel.build.title'
    })), _react.default.createElement("p", null, intl({
      id: 'org.umi.ui.dashboard.panel.build.desc'
    })))),
    body: _react.default.createElement("div", {
      onClick: function onClick() {
        return redirect('/tasks?active=build');
      }
    }, intl({
      id: 'org.umi.ui.dashboard.panel.goto.task'
    }))
  }, {
    className: actionCardCls,
    title: _react.default.createElement("div", {
      className: _uiModule.default.future
    }, _react.default.createElement("div", null, _react.default.createElement(_icons.Smile, null)), _react.default.createElement("p", null, intl({
      id: 'org.umi.ui.dashboard.panel.coming.soon'
    })))
  }];

  if (!closed) {
    actionCards.unshift({
      className: welcomeCardCls,
      size: 'small',
      // extra: <Close className={styles.close} onClick={handleClose} />,
      body: _react.default.createElement("div", null, _react.default.createElement("h2", null, "Hi"), _react.default.createElement("p", null, intl({
        id: 'org.umi.ui.dashboard.panel.welcome.title'
      }, {
        name: currentProject.name
      })), _react.default.createElement("div", null, _react.default.createElement("a", {
        href: "https://umijs.org/guide/umi-ui.html",
        target: "_blank",
        rel: "noopener noreferrer"
      }, basicUI.name || 'Umi', " UI"), intl({
        id: 'org.umi.ui.dashboard.panel.welcome.desc'
      })))
    });
  }

  return _react.default.createElement("div", {
    className: _uiModule.default.container
  }, _react.default.createElement(_antd.Row, {
    className: _uiModule.default['container-row'],
    type: "flex",
    gutter: 24
  }, actionCards.map(function (card, i) {
    var className = card.className,
        body = card.body,
        restProps = _objectWithoutProperties(card, ["className", "body"]);

    return _react.default.createElement(_antd.Col, {
      key: i.toString(),
      className: className,
      xs: 12,
      sm: 12,
      md: 12,
      lg: 12,
      xl: 6
    }, _react.default.createElement(_antd.Card, _extends({
      className: _uiModule.default.card,
      bordered: false,
      hoverable: false,
      children: body
    }, restProps)));
  })));
};

var _default = DashboardUI;
exports.default = _default;
