// umi-ui-tasks/lib/ui/components/Analyze/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _Treemap = _interopRequireDefault(require("./Treemap"));

var _Tooltip = _interopRequireDefault(require("./Tooltip"));

var _filesize = _interopRequireDefault(require("filesize"));

var _indexModule = _interopRequireDefault(require("./index.module.less"));

var _icons = require("@ant-design/icons");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var SIZE_SWITCH_ITEMS = [{
  label: 'Stat',
  prop: 'statSize'
}, {
  label: 'Parsed',
  prop: 'parsedSize'
}, {
  label: 'Gzipped',
  prop: 'gzipSize'
}];

var AnalyzeComponent =
/*#__PURE__*/
function (_Component) {
  _inherits(AnalyzeComponent, _Component);

  function AnalyzeComponent(props) {
    var _this;

    _classCallCheck(this, AnalyzeComponent);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(AnalyzeComponent).call(this, props));
    _this.treemap = null;
    _this.iframeRef = null;

    _this.saveRef = function (treemap) {
      _this.treemap = treemap;
    };

    _this.state = {
      showTooltip: false,
      tooltipContent: '',
      showReloadFlag: false
    };

    _this.handleMouseLeaveTreemap = function () {
      _this.setState({
        showTooltip: false
      });
    };

    _this.getTooltipContent = function (module) {
      if (!module) return null;
      return _react.default.createElement("div", null, _react.default.createElement("div", null, _react.default.createElement("strong", null, module.label)), _react.default.createElement("br", null), _this.renderModuleSize(module, 'stat'), !module.inaccurateSizes && _this.renderModuleSize(module, 'parsed'), !module.inaccurateSizes && _this.renderModuleSize(module, 'gzip'), module.path && _react.default.createElement("div", null, "Path: ", _react.default.createElement("strong", null, module.path)), module.isAsset && _react.default.createElement("div", null, _react.default.createElement("br", null), _react.default.createElement("strong", null, _react.default.createElement("em", null, "Right-click to view options related to this chunk"))));
    };

    _this.handleTreemapGroupHover = function (event) {
      var group = event.group;

      if (!group) {
        _this.setState({
          showTooltip: false
        });

        return;
      }

      _this.setState({
        showTooltip: true,
        tooltipContent: _this.getTooltipContent(group)
      });
    };

    _this.showReload = function () {
      _this.setState({
        showReloadFlag: true
      });
    };

    _this.hideReload = function () {
      _this.setState({
        showReloadFlag: false
      });
    };

    _this.reloadIframe = function () {
      if (!_this.iframeRef || !_this.iframeRef.current) {
        return;
      }

      _this.iframeRef.current.contentWindow.location = _this.props.src;
    };

    _this.iframeRef = _react.default.createRef();
    return _this;
  }

  _createClass(AnalyzeComponent, [{
    key: "renderModuleSize",
    value: function renderModuleSize(module, sizeType) {
      var sizeProp = "".concat(sizeType, "Size");
      var size = module[sizeProp];
      var sizeLabel = SIZE_SWITCH_ITEMS.find(function (item) {
        return item.prop === sizeProp;
      }).label;
      var isActive = true;
      return typeof size === 'number' ? _react.default.createElement("div", {
        className: isActive ? _indexModule.default.activeSize : ''
      }, sizeLabel, " size: ", _react.default.createElement("strong", null, (0, _filesize.default)(size))) : null;
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          analyze = _this$props.analyze,
          src = _this$props.src,
          api = _this$props.api;
      var showReloadFlag = this.state.showReloadFlag;
      var intl = api.intl;

      if (!src && !analyze) {
        return _react.default.createElement("div", {
          className: _indexModule.default.empty
        }, _react.default.createElement("div", null, _react.default.createElement("div", {
          className: "icon"
        }), _react.default.createElement("span", null, intl({
          id: 'org.umi.ui.tasks.analyze.empty'
        }))));
      }

      return _react.default.createElement("div", {
        className: _indexModule.default.container,
        onMouseEnter: this.showReload,
        onMouseLeave: this.hideReload
      }, src ? _react.default.createElement("div", {
        className: _indexModule.default.iframeContainer
      }, showReloadFlag ? _react.default.createElement("div", {
        className: _indexModule.default.reload,
        onClick: this.reloadIframe
      }, _react.default.createElement(_icons.Reload, null)) : null, _react.default.createElement("iframe", {
        ref: this.iframeRef,
        className: _indexModule.default.iframe,
        src: src
      })) : _react.default.createElement(_react.default.Fragment, null, _react.default.createElement(_Treemap.default, {
        className: _indexModule.default.map,
        ref: this.saveRef,
        weightProp: "parsedSize",
        data: analyze.visibleChunks,
        onMouseLeave: this.handleMouseLeaveTreemap,
        onGroupHover: this.handleTreemapGroupHover
      }), _react.default.createElement(_Tooltip.default, {
        visible: this.state.showTooltip
      }, this.state.tooltipContent)));
    }
  }]);

  return AnalyzeComponent;
}(_react.Component);

var _default = AnalyzeComponent;
exports.default = _default;
