// af-webpack/lib/getUserConfig/configs/minimizer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = _default;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function _react() {
    return data;
  };

  return data;
}

function _assert() {
  const data = _interopRequireDefault(require("assert"));

  _assert = function _assert() {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _default() {
  return {
    name: 'minimizer',

    validate(val) {
      (0, _assert().default)(val === 'terserjs' || val === 'uglifyjs', `minimizer should be terserjs or uglifyjs, but got ${val}`);
    },

    group: 'webpack',
    default: 'uglifyjs',
    type: 'list',
    choices: ['terserjs', 'uglifyjs'],
    title: {
      'zh-CN': '压缩器',
      'en-US': 'Minimizer'
    },
    description: {
      'zh-CN': '使用 terserjs 压缩时有 es6 语法不会报错。',
      'en-US': 'There is no error with es6 syntax when compressing with terserjs.'
    }
  };
}
