// af-webpack/lib/getUserConfig/configs/publicPath.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = _default;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function _react() {
    return data;
  };

  return data;
}

function _assert() {
  const data = _interopRequireDefault(require("assert"));

  _assert = function _assert() {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _default() {
  return {
    name: 'publicPath',

    validate(val) {
      (0, _assert().default)(typeof val === 'string', `The publicPath config must be String, but got ${val}.`);
      (0, _assert().default)(val.endsWith('/'), `The publicPath config must ends with /.`);
    },

    default: '/',
    title: 'publicPath',
    type: 'string',
    group: 'deploy',
    description: {
      'zh-CN': '指定 webpack 的 publicPath 配置，指向发布后的静态资源位置，dev 时无效。',
      'en-US': 'The publicPath of your assets.'
    }
  };
}
