// umi-ui-tasks/lib/ui/components/Terminal/index.js
"use strict";

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _antd = require("antd");

var _icons = require("@ant-design/icons");

var _react = _interopRequireWildcard(require("react"));

var _hooks = require("../../hooks");

var _indexModule = _interopRequireDefault(require("./index.module.less"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { if (!(Symbol.iterator in Object(arr) || Object.prototype.toString.call(arr) === "[object Arguments]")) { return; } var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var TerminalComponent = function TerminalComponent(_ref) {
  var terminal = _ref.terminal,
      log = _ref.log,
      onClear = _ref.onClear,
      _ref$size = _ref.size,
      size = _ref$size === void 0 ? {} : _ref$size,
      api = _ref.api;
  var intl = api.intl;
  var domContainer = (0, _react.useRef)();

  var _useState = (0, _react.useState)(false),
      _useState2 = _slicedToArray(_useState, 2),
      setInit = _useState2[1];

  (0, _react.useEffect)(function () {
    setInit(true);
  }, []);
  (0, _hooks.useTerminal)({
    terminal: terminal,
    ref: domContainer.current
  });
  (0, _react.useEffect)(function () {
    if (log) {
      terminal.write(log.replace(/\n/g, '\r\n'));
    }

    return function () {
      terminal.clear();
    };
  }, [log]); // mount or updated

  var prevSize = (0, _hooks.usePrevious)(size);
  (0, _react.useEffect)(function () {
    if (prevSize) {
      if (prevSize.width !== size.width || prevSize.height !== size.height) {
        terminal.fit();
      }
    }
  }, [size.width, size.height]);

  var clear = function clear() {
    terminal.clear();
    onClear && onClear();
  };

  var toBottom = function toBottom() {
    terminal.scrollToBottom();
  };

  return _react.default.createElement("div", {
    className: _indexModule.default.wrapper
  }, _react.default.createElement(_antd.Row, {
    className: _indexModule.default.titleWrapper
  }, _react.default.createElement(_antd.Col, {
    span: 8,
    className: _indexModule.default.formmatGroup
  }, intl({
    id: 'org.umi.ui.tasks.log.title'
  })), _react.default.createElement(_antd.Col, {
    span: 4,
    offset: 12,
    className: _indexModule.default.actionGroup
  }, _react.default.createElement("span", {
    className: _indexModule.default.icon
  }, _react.default.createElement(_antd.Tooltip, {
    title: intl({
      id: 'org.umi.ui.tasks.terminal.clear'
    })
  }, _react.default.createElement(_icons.Delete, {
    onClick: clear
  }))), _react.default.createElement("span", {
    className: _indexModule.default.icon
  }, _react.default.createElement(_antd.Tooltip, {
    title: intl({
      id: 'org.umi.ui.tasks.terminal.bottom'
    })
  }, _react.default.createElement(_icons.Enter, {
    onClick: toBottom
  }))))), _react.default.createElement("div", {
    ref: domContainer,
    className: _indexModule.default.logContainer
  }));
};

var _default = TerminalComponent;
exports.default = _default;
