// umi-build-dev/lib/plugins/404/NotFound.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _link = _interopRequireDefault(require("umi/link"));

var _withRouter = _interopRequireDefault(require("umi/withRouter"));

require("whatwg-fetch");

var _guessJSFileFromPath = _interopRequireDefault(require("./guessJSFileFromPath"));

var _NotFound = _interopRequireDefault(require("./NotFound.less"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var NotFound =
/*#__PURE__*/
function (_React$Component) {
  _inherits(NotFound, _React$Component);

  function NotFound(props) {
    var _this;

    _classCallCheck(this, NotFound);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(NotFound).call(this, props));
    _this.state = {
      loading: true,
      routes: []
    };
    return _this;
  }

  _createClass(NotFound, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      var _this2 = this;

      fetch('/__umiDev/routes').then(function (res) {
        return res.json();
      }).then(function (routes) {
        _this2.setState({
          loading: false,
          routes: routes
        });
      });
    }
  }, {
    key: "renderRoutes",
    value: function renderRoutes(routes) {
      var _this3 = this;

      return _react.default.createElement("ul", null, routes.map(function (route, i) {
        if (!route.path) return null;
        return _react.default.createElement("li", {
          key: route.key || i
        }, _react.default.createElement(_link.default, {
          to: route.path
        }, route.path), route.routes ? _this3.renderRoutes(route.routes) : null);
      }));
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          location = _this$props.location,
          pagesPath = _this$props.pagesPath,
          hasRoutesInConfig = _this$props.hasRoutesInConfig;
      var jsFile = (0, _guessJSFileFromPath.default)(location.pathname);
      return _react.default.createElement("div", {
        className: _NotFound.default['umi-NotFound-wrapper']
      }, _react.default.createElement("h1", null, "umi development 404 page"), _react.default.createElement("p", null, "There's not a page yet at ", _react.default.createElement("code", null, location.pathname), "."), _react.default.createElement("p", null, "Create a React.js component in your pages directory at", ' ', _react.default.createElement("code", null, pagesPath, "/", jsFile), ' ', hasRoutesInConfig ? "and configure the route in config file " : '', "then this page will automatically refresh to show the new page component you created."), _react.default.createElement("h2", null, "Your Routes"), this.state.loading ? _react.default.createElement("div", null, "Loading routes...") : this.renderRoutes(this.state.routes));
    }
  }]);

  return NotFound;
}(_react.default.Component);

var _default = (0, _withRouter.default)(NotFound);

exports.default = _default;
