// umi-plugin-ui/lib/plugins/dashboard/ui.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _antd = require("antd");

var _index = _interopRequireDefault(require("./ui/index"));

var _zhCN = _interopRequireDefault(require("./locales/zh-CN"));

var _enUS = _interopRequireDefault(require("./locales/en-US"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = function _default(api) {
  var intl = api.intl;
  var FormattedMessage = intl.FormattedMessage;

  var ConfigAction = function ConfigAction(_ref) {
    var api = _ref.api;
    return _react.default.createElement(_antd.Button, {
      onClick: function onClick() {
        api.launchEditor({
          type: 'project'
        });
      },
      type: "default"
    }, _react.default.createElement(FormattedMessage, {
      id: "org.umi.ui.dashboard.launch.editor"
    }));
  };

  api.addLocales({
    'zh-CN': _zhCN.default,
    'en-US': _enUS.default
  });
  api.addPanel({
    title: 'org.umi.ui.dashboard.panel',
    path: '/dashboard',
    actions: api.mini ? [] : [_react.default.createElement(ConfigAction, {
      api: api
    })],
    icon: {
      type: 'dashboard',
      theme: 'filled'
    },
    component: function component() {
      return _react.default.createElement(_index.default, {
        api: api
      });
    }
  });
};

exports.default = _default;
