// umi-build-dev/lib/plugins/commands/block/ui/client/TagSelect/index.js
"use strict";

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _antd = require("antd");

var _icons = require("@ant-design/icons");

var _indexModule = _interopRequireDefault(require("./index.module.less"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { if (!(Symbol.iterator in Object(arr) || Object.prototype.toString.call(arr) === "[object Arguments]")) { return; } var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var CheckableTag = _antd.Tag.CheckableTag;
var sortTagMap = {
  表格: 10,
  表单: 9,
  通用: 8,
  列表: 7,
  用户: 6,
  数据录入: 5,
  数据展示: 4,
  基本: 3,
  高级: 2
};

var sortTag = function sortTag(a, b) {
  if (sortTagMap[a]) {
    return -sortTagMap[a];
  }

  if (sortTagMap[b]) {
    return -sortTagMap[b];
  }

  return 0;
};

var TagSelect = function TagSelect(_ref) {
  var value = _ref.value,
      tagList = _ref.tagList,
      _onChange = _ref.onChange,
      loading = _ref.loading,
      expandText = _ref.expandText,
      allText = _ref.allText,
      collapseText = _ref.collapseText;

  var _useState = (0, _react.useState)(true),
      _useState2 = _slicedToArray(_useState, 2),
      expand = _useState2[0],
      setExpandValue = _useState2[1];

  var _useState3 = (0, _react.useState)(false),
      _useState4 = _slicedToArray(_useState3, 2),
      hasExpandButton = _useState4[0],
      setHasExpandButton = _useState4[1];

  var ref = (0, _react.useRef)(undefined);
  (0, _react.useLayoutEffect)(function () {
    // 计算需不需要折行
    if (ref && ref.current) {
      var clientHeight = ref.current.clientHeight;

      if (clientHeight > 30) {
        setHasExpandButton(true);
      } else {
        setHasExpandButton(false);
      }
    }
  }, [loading, tagList.length]);
  return _react.default.createElement("div", {
    className: "".concat(_indexModule.default.tagContainer, " ").concat(expand && hasExpandButton && _indexModule.default.expand, " ").concat(hasExpandButton && _indexModule.default.hasExpandButton)
  }, _react.default.createElement("div", {
    className: _indexModule.default.tagView,
    ref: ref
  }, !loading && _react.default.createElement(_react.default.Fragment, null, _react.default.createElement(CheckableTag, {
    checked: value === '',
    onChange: function onChange() {
      _onChange('');
    }
  }, allText), _toConsumableArray(tagList).sort(sortTag).filter(function (tagName) {
    return tagName !== '废弃';
  }).map(function (tag) {
    return _react.default.createElement(CheckableTag, {
      checked: value === tag,
      key: tag,
      onChange: function onChange(checked) {
        if (checked) {
          _onChange(tag);
        } else {
          _onChange('');
        }
      }
    }, tag);
  }), _react.default.createElement("span", {
    style: {
      flex: 999
    }
  })), loading && _react.default.createElement(_icons.Loading, null)), !loading && hasExpandButton ? _react.default.createElement("a", {
    className: _indexModule.default.expandButton,
    onClick: function onClick() {
      setExpandValue(!expand);
    }
  }, _react.default.createElement(_react.default.Fragment, null, expand ? expandText : collapseText, " ", _react.default.createElement(_icons.Up, {
    className: _indexModule.default.upIcon
  }))) : '');
};

var _default = TagSelect;
exports.default = _default;
