// bizcharts/es6/components/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Facet = exports.Label = exports.Guide = exports.View = exports.Geom = exports.Tooltip = exports.Legend = exports.Axis = exports.Coord = exports.Chart = undefined;

var _Chart2 = require('./Chart');

var _Chart3 = _interopRequireDefault(_Chart2);

var _Coord2 = require('./Coord');

var _Coord3 = _interopRequireDefault(_Coord2);

var _Axis2 = require('./Axis');

var _Axis3 = _interopRequireDefault(_Axis2);

var _Legend2 = require('./Legend');

var _Legend3 = _interopRequireDefault(_Legend2);

var _Tooltip2 = require('./Tooltip');

var _Tooltip3 = _interopRequireDefault(_Tooltip2);

var _Geom2 = require('./Geom');

var _Geom3 = _interopRequireDefault(_Geom2);

var _View2 = require('./View');

var _View3 = _interopRequireDefault(_View2);

var _Guide2 = require('./Guide');

var _Guide3 = _interopRequireDefault(_Guide2);

var _Label2 = require('./Label');

var _Label3 = _interopRequireDefault(_Label2);

var _Facet2 = require('./Facet');

var _Facet3 = _interopRequireDefault(_Facet2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.Chart = _Chart3.default;
exports.Coord = _Coord3.default;
exports.Axis = _Axis3.default;
exports.Legend = _Legend3.default;
exports.Tooltip = _Tooltip3.default;
exports.Geom = _Geom3.default;
exports.View = _View3.default;
exports.Guide = _Guide3.default;
exports.Label = _Label3.default;
exports.Facet = _Facet3.default;
