// umi-build-dev/lib/plugins/commands/block/ui/server/core/tasks/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "parseUrl", {
  enumerable: true,
  get: function get() {
    return _parse_url.default;
  }
});
Object.defineProperty(exports, "gitClone", {
  enumerable: true,
  get: function get() {
    return _git_clone.default;
  }
});
Object.defineProperty(exports, "gitUpdate", {
  enumerable: true,
  get: function get() {
    return _git_update.default;
  }
});
Object.defineProperty(exports, "runGenerator", {
  enumerable: true,
  get: function get() {
    return _run_generator.default;
  }
});
Object.defineProperty(exports, "writeRoutes", {
  enumerable: true,
  get: function get() {
    return _write_routes.default;
  }
});
Object.defineProperty(exports, "install", {
  enumerable: true,
  get: function get() {
    return _install.default;
  }
});

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function _react() {
    return data;
  };

  return data;
}

var _parse_url = _interopRequireDefault(require("./parse_url"));

var _git_clone = _interopRequireDefault(require("./git_clone"));

var _git_update = _interopRequireDefault(require("./git_update"));

var _run_generator = _interopRequireDefault(require("./run_generator"));

var _write_routes = _interopRequireDefault(require("./write_routes"));

var _install = _interopRequireDefault(require("./install"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
