// umi-build-dev/lib/plugins/commands/block/ui/client/GlobalSearch.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _antd = require("antd");

var _icons = require("@ant-design/icons");

var _GlobalSearchModule = _interopRequireDefault(require("./GlobalSearch.module.less"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var GlobalSearch = function GlobalSearch(props) {
  var onChange = props.onChange,
      api = props.api;
  var intl = api.intl,
      _ = api._; // 时间太长会造成卡顿的感觉，200-300 比较合适

  var handleChangeDebounce = _.debounce(function (value) {
    onChange(value);
  }, 300);

  return _react.default.createElement(_antd.Input, {
    prefix: _react.default.createElement(_icons.Search, null),
    className: _GlobalSearchModule.default.search,
    allowClear: true,
    onChange: function onChange(e) {
      return handleChangeDebounce(e.target.value);
    },
    placeholder: intl({
      id: 'org.umi.ui.blocks.content.search_block'
    })
  });
};

var _default = GlobalSearch;
exports.default = _default;
