// af-webpack/lib/getWebpackConfig/eslint.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = _default;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function _react() {
    return data;
  };

  return data;
}

function _eslintFormatter() {
  const data = _interopRequireDefault(require("react-dev-utils/eslintFormatter"));

  _eslintFormatter = function _eslintFormatter() {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _default(webpackConfig, opts) {
  const eslintOptions = {
    formatter: _eslintFormatter().default,
    baseConfig: {
      extends: [require.resolve('eslint-config-umi')]
    },
    ignore: false,
    eslintPath: require.resolve('eslint'),
    useEslintrc: false
  };
  webpackConfig.module.rule('eslint').test(/\.(js|jsx)$/).include.add(opts.cwd).end().exclude.add(/node_modules/).end().enforce('pre').use('eslint-loader').loader(require.resolve('eslint-loader')).options(eslintOptions);
}
