// umi-plugin-react/lib/plugins/pwa/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = _default;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function _react() {
    return data;
  };

  return data;
}

function _path() {
  const data = require("path");

  _path = function _path() {
    return data;
  };

  return data;
}

function _assert() {
  const data = _interopRequireDefault(require("assert"));

  _assert = function _assert() {
    return data;
  };

  return data;
}

function _chalk() {
  const data = _interopRequireDefault(require("chalk"));

  _chalk = function _chalk() {
    return data;
  };

  return data;
}

function _workboxWebpackPlugin() {
  const data = _interopRequireDefault(require("workbox-webpack-plugin"));

  _workboxWebpackPlugin = function _workboxWebpackPlugin() {
    return data;
  };

  return data;
}

var _WebManifestPlugin = _interopRequireDefault(require("./WebManifestPlugin"));

var _generateWebManifest2 = _interopRequireDefault(require("./generateWebManifest"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _default(api, options) {
  const pkg = api.pkg,
        relativeToTmp = api.relativeToTmp,
        publicPath = api.config.publicPath,
        absSrcPath = api.paths.absSrcPath;
  (0, _assert().default)(pkg && pkg.name, `You must have ${_chalk().default.underline.cyan('package.json')} and configure ${_chalk().default.underline.cyan('name')} in it when enable pwa.`); // generate webmanifest before workbox generation, so that webmanifest can be added to precached list

  const _generateWebManifest = (0, _generateWebManifest2.default)(api, _objectSpread({}, options.manifestOptions)),
        srcPath = _generateWebManifest.srcPath,
        outputPath = _generateWebManifest.outputPath;

  api.chainWebpackConfig(webpackConfig => {
    webpackConfig.plugin('webmanifest').use(_WebManifestPlugin.default, [{
      publicPath,
      srcPath,
      outputPath,
      pkgName: pkg.name
    }]);
  });

  if (process.env.NODE_ENV === 'production') {
    const mode = options.workboxPluginMode || 'GenerateSW';
    const defaultGenerateSWOptions = mode === 'GenerateSW' ? {
      cacheId: pkg.name,
      skipWaiting: true,
      clientsClaim: true
    } : {
      swSrc: (0, _path().join)(absSrcPath, 'service-worker.js')
    };

    const workboxConfig = _objectSpread({
      // remove manifest.json from exclude list. https://github.com/GoogleChrome/workbox/issues/1665
      exclude: [/\.map$/, /favicon\.ico$/, /^manifest.*\.js?$/]
    }, defaultGenerateSWOptions, {}, options.workboxOptions || {});

    const swDest = workboxConfig.swDest || workboxConfig.swSrc && (0, _path().basename)(workboxConfig.swSrc) || 'service-worker.js';
    api.chainWebpackConfig(webpackConfig => {
      webpackConfig.plugin('workbox').use(_workboxWebpackPlugin().default[mode], [workboxConfig]);
      webpackConfig.resolve.alias.set('register-service-worker', require.resolve('register-service-worker'));
    });
    api.addEntryCode(`
var registerSW = require('${relativeToTmp((0, _path().join)(__dirname, './registerServiceWorker.js'))}').default;
registerSW('${swDest}');
    `.trim());
  }
}
