// umi-build-dev/lib/plugins/commands/block/ui/server/core/Flow.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function _react() {
    return data;
  };

  return data;
}

function _events() {
  const data = require("events");

  _events = function _events() {
    return data;
  };

  return data;
}

var _enum = require("./enum");

var _Logger = _interopRequireDefault(require("./Logger"));

var _exec = _interopRequireDefault(require("../util/exec"));

var _tasks = require("./tasks");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

class Flow extends _events().EventEmitter {
  // ç”¨æˆ·å–æ¶ˆ
  constructor({
    api
  }) {
    super();
    this.api = void 0;
    this.ctx = void 0;
    this.tasks = [];
    this.isCancel = false;
    this.logger = void 0;
    this.proc = void 0;
    this.state = _enum.FlowState.INIT;
    this.api = api;
    this.logger = new _Logger.default();
    this.logger.on('log', data => {
      this.emit('log', data);
    });
    this.ctx = {
      execa: (0, _exec.default)(this.logger, this.setProcRef.bind(this)),
      api: this.api,
      logger: this.logger,
      stages: {},
      result: {}
    };
    this.registryTasks();
  }

  run(args) {
    var _this = this;

    return _asyncToGenerator(function* () {
      _this.state = _enum.FlowState.ING;
      let hasBreak = false;
      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = _this.tasks[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          const _step$value = _step.value,
                name = _step$value.name,
                task = _step$value.task,
                state = _step$value.state;

          // ç”¨æˆ·å–æ¶ˆä»»åŠ¡
          if (_this.isCancel) {
            hasBreak = true;

            _this.setStepState(name, _enum.StepState.CANCEL);

            break;
          }

          if (state === _enum.StepState.SUCCESS) {
            continue;
          }

          try {
            yield task(_this.ctx, args);

            _this.setStepState(name, _enum.StepState.SUCCESS);
          } catch (e) {
            hasBreak = true;
            /**
             * æŠ›é”™æœ‰ä¸¤ç§æƒ…å†µ
             *  1. ä»»åŠ¡æ‰§è¡Œå‡ºé”™
             *  2. ç”¨æˆ·å–æ¶ˆåŽï¼Œä¼šæ€æ­»å­è¿›ç¨‹ï¼Œå­è¿›ç¨‹å¯èƒ½ edit(1)
             */

            if (!_this.isCancel) {
              _this.state = _enum.FlowState.FAIL;

              _this.setStepState(name, _enum.StepState.FAIL);

              _this.emit('state', _objectSpread({}, args, {
                state: _enum.FlowState.FAIL
              }));

              _this.emit('log', {
                data: `\nðŸš§  Execute task error: ${e.message}\n`
              });
            }

            break;
          }
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator.return != null) {
            _iterator.return();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }

      if (hasBreak) {
        return _this.ctx.result;
      } // å®Œæˆä¹‹åŽè§¦å‘ä¸€ä¸‹å®Œæˆäº‹ä»¶ï¼Œå‰ç«¯æ›´æ–°ä¸€ä¸‹æŒ‰é’®çŠ¶æ€


      _this.state = _enum.FlowState.SUCCESS;
      const generator = _this.ctx.stages.generator;

      _this.emit('state', {
        data: _objectSpread({}, args, {
          previewUrl: `http://localhost:${process.env.PORT || '8000'}${generator.path.toLowerCase()}`
        }),
        state: _enum.FlowState.SUCCESS
      }); // æ¸…ç©ºæ—¥å¿—


      _this.logger.clear();

      return _this.ctx.result;
    })();
  }

  cancel() {
    if (this.state !== _enum.FlowState.ING) {
      const err = new Error(`Error state(${this.state}) to terminated`);
      err.name = 'FlowError';
      throw err;
    }

    this.isCancel = true;
    this.state = _enum.FlowState.CANCEL;

    if (this.proc) {
      this.proc.kill('SIGTERM');
    }

    setTimeout(() => {
      this.emit('log', {
        data: '\nðŸ›‘  Stopped task success!\n'
      });
    }, 2000);
  }
  /**
   * é‡è¯•
   */


  retry(args) {
    var _this2 = this;

    return _asyncToGenerator(function* () {
      if (_this2.state !== _enum.FlowState.FAIL) {
        const err = new Error(`Error state(${_this2.state}) to retry`);
        err.name = 'FlowError';
        throw err;
      }

      return _this2.run(args);
    })();
  }

  getLog() {
    return this.logger.getLog();
  }

  getBlockUrl() {
    if (this.state !== _enum.FlowState.ING) {
      return '';
    }

    return this.ctx.result.blockUrl;
  }

  setStepState(taskName, state) {
    const curTask = this.tasks.find(({
      name
    }) => name === taskName);

    if (!curTask) {
      return;
    }

    curTask.state = state;
  }

  registryTasks() {
    [{
      name: 'parseUrl',
      task: _tasks.parseUrl
    }, {
      name: 'gitClone',
      task: _tasks.gitClone
    }, {
      name: 'gitUpdate',
      task: _tasks.gitUpdate
    }, {
      name: 'install',
      task: _tasks.install
    }, {
      name: 'runGenerator',
      task: _tasks.runGenerator
    }, {
      name: 'writeRoutes',
      task: _tasks.writeRoutes
    }].forEach(({
      name,
      task
    }) => {
      this.tasks.push({
        name,
        task,
        state: _enum.StepState.INIT
      });
    });
  }

  setProcRef(proc) {
    this.proc = proc;
  }

}

var _default = Flow;
exports.default = _default;
