// umi-build-dev/lib/plugins/commands/block/ui/client/Adder/AddBlockForm.js
"use strict";

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _antd = require("antd");

var _UIApiContext = _interopRequireDefault(require("../UIApiContext"));

var _InfoToolTip = _interopRequireDefault(require("./InfoToolTip"));

var _RouteFilesTreeData = _interopRequireDefault(require("./RouteFilesTreeData"));

var _BlockItem = require("../BlockList/BlockItem");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 *  在 umi ui 中添加区块
 *  需要选择安装的具体文件位置
 */
var AddBlockFormForUI = function AddBlockFormForUI(_ref) {
  var form = _ref.form,
      visible = _ref.visible;

  var _useContext = (0, _react.useContext)(_UIApiContext.default),
      api = _useContext.api;

  var intl = api.intl;
  return _react.default.createElement(_react.default.Fragment, null, _react.default.createElement("div", {
    style: {
      display: 'flex',
      marginBottom: 24
    }
  }, intl({
    id: 'org.umi.ui.blocks.adder.minirecommend'
  })), _react.default.createElement(_antd.Form.Item, {
    name: "path",
    label: _react.default.createElement(_InfoToolTip.default, {
      title: intl({
        id: 'org.umi.ui.blocks.adder.path'
      }),
      placeholder: intl({
        id: 'org.umi.ui.blocks.adder.path.tooltip'
      })
    }),
    rules: [{
      required: true,
      message: intl({
        id: 'org.umi.ui.blocks.adder.path.rule'
      })
    }]
  }, _react.default.createElement(_RouteFilesTreeData.default, {
    visible: visible
  })), _react.default.createElement(_antd.Form.Item, {
    name: "name",
    label: _react.default.createElement(_InfoToolTip.default, {
      title: intl({
        id: 'org.umi.ui.blocks.adder.name'
      }),
      placeholder: intl({
        id: 'org.umi.ui.blocks.adder.name.tooltip'
      })
    }),
    rules: [{
      validator: function validator(rule, name) {
        var filePath, _ref2, exists, blockFileTarget, _ref3, varExists;

        return regeneratorRuntime.async(function validator$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (name) {
                  _context.next = 2;
                  break;
                }

                throw new Error(intl({
                  id: 'org.umi.ui.blocks.adder.name.required'
                }));

              case 2:
                if (/^[a-zA-Z$_][a-zA-Z\d_]*$/.test(name)) {
                  _context.next = 4;
                  break;
                }

                throw new Error(intl({
                  id: 'org.umi.ui.blocks.adder.name.illegal'
                }));

              case 4:
                if (/^(?:[A-Z][a-z]+)+$/.test(name)) {
                  _context.next = 6;
                  break;
                }

                throw new Error(intl({
                  id: 'org.umi.ui.blocks.adder.name.illegalReact'
                }));

              case 6:
                _context.next = 8;
                return regeneratorRuntime.awrap((0, _BlockItem.getPathFromFilename)(api, form.getFieldValue('path')));

              case 8:
                filePath = _context.sent;
                _context.next = 11;
                return regeneratorRuntime.awrap(api.callRemote({
                  type: 'org.umi.block.checkExistFilePath',
                  payload: {
                    path: "".concat(filePath, "/").concat(name)
                  }
                }));

              case 11:
                _ref2 = _context.sent;
                exists = _ref2.exists;

                if (!exists) {
                  _context.next = 15;
                  break;
                }

                throw new Error(intl({
                  id: 'org.umi.ui.blocks.adder.pathexist'
                }));

              case 15:
                blockFileTarget = form.getFieldValue('path');
                _context.next = 18;
                return regeneratorRuntime.awrap(api.callRemote({
                  type: 'org.umi.block.checkBindingInFile',
                  payload: {
                    path: blockFileTarget,
                    name: name
                  }
                }));

              case 18:
                _ref3 = _context.sent;
                varExists = _ref3.exists;

                if (!varExists) {
                  _context.next = 22;
                  break;
                }

                throw new Error(intl({
                  id: 'org.umi.ui.blocks.adder.varexist'
                }));

              case 22:
              case "end":
                return _context.stop();
            }
          }
        });
      }
    }]
  }, _react.default.createElement(_antd.Input, {
    placeholder: intl({
      id: 'org.umi.ui.blocks.adder.name.placeholder'
    })
  })));
};

var _default = AddBlockFormForUI;
exports.default = _default;
