// umi-build-dev/lib/plugins/commands/block/ui/client/Adder/AddBlockFormForUI.js
"use strict";

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _antd = require("antd");

var _UIApiContext = _interopRequireDefault(require("../UIApiContext"));

var _InfoToolTip = _interopRequireDefault(require("./InfoToolTip"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { if (!(Symbol.iterator in Object(arr) || Object.prototype.toString.call(arr) === "[object Arguments]")) { return; } var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var PathLabel = function PathLabel(_ref) {
  var value = _ref.value,
      name = _ref.name;
  return _react.default.createElement("div", {
    style: {
      display: 'flex',
      marginBottom: 24
    }
  }, "\u533A\u5757\u5C06\u88AB\u5B89\u88C5\u5230", _react.default.createElement("code", {
    style: {
      backgroundColor: '#3b3b4d',
      margin: '0 8px',
      padding: '0 8px',
      borderRadius: 4
    }
  }, "".concat(value).concat(name ? "/".concat(name) : '').replace(/\//g, '/').replace(/\/\//g, '/')));
};

var AddBlockFormForUI = function AddBlockFormForUI(_ref2) {
  var blockTarget = _ref2.blockTarget,
      form = _ref2.form;

  var _useContext = (0, _react.useContext)(_UIApiContext.default),
      api = _useContext.api;

  var intl = api.intl;

  var _useState = (0, _react.useState)(form.getFieldValue('name')),
      _useState2 = _slicedToArray(_useState, 2),
      varName = _useState2[0],
      setValueName = _useState2[1];

  return _react.default.createElement(_react.default.Fragment, null, _react.default.createElement(_antd.Form.Item, {
    noStyle: true,
    name: "path",
    label: _react.default.createElement(_InfoToolTip.default, {
      title: intl({
        id: 'org.umi.ui.blocks.adder.path'
      }),
      placeholder: intl({
        id: 'org.umi.ui.blocks.adder.path.minitooltip'
      })
    }),
    rules: [{
      required: true,
      message: intl({
        id: 'org.umi.ui.blocks.adder.path.rule'
      })
    }]
  }, _react.default.createElement(PathLabel, {
    name: varName
  })), _react.default.createElement(_antd.Form.Item, {
    name: "name",
    label: _react.default.createElement(_InfoToolTip.default, {
      title: intl({
        id: 'org.umi.ui.blocks.adder.name'
      }),
      placeholder: intl({
        id: 'org.umi.ui.blocks.adder.name.tooltip'
      })
    }),
    rules: [{
      validator: function validator(rule, name) {
        var _ref3, exists, blockFileTarget, _ref4, varExists;

        return regeneratorRuntime.async(function validator$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (name) {
                  _context.next = 2;
                  break;
                }

                throw new Error(intl({
                  id: 'org.umi.ui.blocks.adder.name.required'
                }));

              case 2:
                if (/^[a-zA-Z$_][a-zA-Z\d_]*$/.test(name)) {
                  _context.next = 4;
                  break;
                }

                throw new Error(intl({
                  id: 'org.umi.ui.blocks.adder.name.illegal'
                }));

              case 4:
                if (/^(?:[A-Z][a-z]+)+$/.test(name)) {
                  _context.next = 6;
                  break;
                }

                throw new Error(intl({
                  id: 'org.umi.ui.blocks.adder.name.illegalReact'
                }));

              case 6:
                _context.next = 8;
                return regeneratorRuntime.awrap(api.callRemote({
                  type: 'org.umi.block.checkExistFilePath',
                  payload: {
                    path: "".concat(blockTarget, "/").concat(name)
                  }
                }));

              case 8:
                _ref3 = _context.sent;
                exists = _ref3.exists;

                if (!exists) {
                  _context.next = 12;
                  break;
                }

                throw new Error(intl({
                  id: 'org.umi.ui.blocks.adder.pathexist'
                }));

              case 12:
                blockFileTarget = form.getFieldValue('path');
                _context.next = 15;
                return regeneratorRuntime.awrap(api.callRemote({
                  type: 'org.umi.block.checkBindingInFile',
                  payload: {
                    path: blockFileTarget,
                    name: name
                  }
                }));

              case 15:
                _ref4 = _context.sent;
                varExists = _ref4.exists;

                if (!varExists) {
                  _context.next = 19;
                  break;
                }

                throw new Error(intl({
                  id: 'org.umi.ui.blocks.adder.varexist'
                }));

              case 19:
              case "end":
                return _context.stop();
            }
          }
        });
      }
    }]
  }, _react.default.createElement(_antd.Input, {
    placeholder: intl({
      id: 'org.umi.ui.blocks.adder.name.placeholder'
    }),
    onChange: function onChange(e) {
      setValueName(e.target.value);
    }
  })));
};

var _default = AddBlockFormForUI;
exports.default = _default;
