// umi-build-dev/lib/plugins/commands/block/ui/client/Adder/index.js
"use strict";

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _antd = require("antd");

var _uppercamelcase = _interopRequireDefault(require("uppercamelcase"));

var _UIApiContext = _interopRequireDefault(require("../UIApiContext"));

var _useCallData2 = _interopRequireDefault(require("../hooks/useCallData"));

var _LogPanel = _interopRequireDefault(require("../LogPanel"));

var _ResultPanel = _interopRequireDefault(require("./ResultPanel"));

var _AddTemplateForm = _interopRequireDefault(require("./AddTemplateForm"));

var _AddBlockFormForUI = _interopRequireDefault(require("./AddBlockFormForUI"));

var _AddBlockForm = _interopRequireDefault(require("./AddBlockForm"));

var _BlockItem = require("../BlockList/BlockItem");

var _util = require("../util");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { if (!(Symbol.iterator in Object(arr) || Object.prototype.toString.call(arr) === "[object Arguments]")) { return; } var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

/**
 * 安装区块
 * @param api
 * @param params
 */
var addBlock = function addBlock(api, params) {
  var _ref, _ref$data, info;

  return regeneratorRuntime.async(function addBlock$(_context) {
    while (1) {
      switch (_context.prev = _context.next) {
        case 0:
          _context.next = 2;
          return regeneratorRuntime.awrap(api.callRemote({
            type: 'org.umi.block.add',
            payload: params
          }));

        case 2:
          _ref = _context.sent;
          _ref$data = _ref.data;
          info = _ref$data === void 0 ? {
            message: ''
          } : _ref$data;
          return _context.abrupt("return", info.message);

        case 6:
        case "end":
          return _context.stop();
      }
    }
  });
};

var cancelAddBlockTask = function cancelAddBlockTask(api) {
  return api.callRemote({
    type: 'org.umi.block.cancel'
  });
};

var Adder = function Adder(props) {
  var visible = props.visible,
      blockTarget = props.blockTarget,
      onAddBlockChange = props.onAddBlockChange,
      onHideModal = props.onHideModal,
      index = props.index,
      _props$block = props.block,
      block = _props$block === void 0 ? {
    url: ''
  } : _props$block,
      blockType = props.blockType;

  var _useContext = (0, _react.useContext)(_UIApiContext.default),
      api = _useContext.api;

  var callRemote = api.callRemote,
      intl = api.intl;

  var _useState = (0, _react.useState)(false),
      _useState2 = _slicedToArray(_useState, 2),
      taskLoading = _useState2[0],
      setTaskLoading = _useState2[1]; // 防止重复提交


  var _useState3 = (0, _react.useState)(false),
      _useState4 = _slicedToArray(_useState3, 2),
      fromCheck = _useState4[0],
      setFromCheck = _useState4[1];

  var _Form$useForm = _antd.Form.useForm(),
      _Form$useForm2 = _slicedToArray(_Form$useForm, 1),
      form = _Form$useForm2[0]; // 展示哪个界面
  // log 日志  form 表单


  var _useState5 = (0, _react.useState)('form'),
      _useState6 = _slicedToArray(_useState5, 2),
      addStatus = _useState6[0],
      setAddStatus = _useState6[1]; // 预览界面需要消费的日志


  var _useState7 = (0, _react.useState)(undefined),
      _useState8 = _slicedToArray(_useState7, 2),
      succeededBlock = _useState8[0],
      setSucceededBlock = _useState8[1];

  var _useCallData = (0, _useCallData2.default)(function () {
    return callRemote({
      type: '@@project/getNpmClients'
    });
  }, [], {
    defaultData: ['npm']
  }),
      _useCallData$data = _useCallData.data,
      npmClients = _useCallData$data === void 0 ? [] : _useCallData$data;

  (0, _react.useEffect)(function () {
    if (api.detectNpmClients) {
      api.detectNpmClients().then(function (clients) {
        form.setFieldsValue({
          npmClient: clients.find(function (c) {
            return npmClients.includes(c);
          })
        });
      });
    }
  }, [npmClients]);
  (0, _react.useEffect)(function () {
    /**
     * 成功之后清理状态
     */
    api.listenRemote({
      type: 'org.umi.block.add-blocks-success',
      onMessage: function onMessage(msg) {
        setTaskLoading(false);
        onAddBlockChange(undefined); // 如果标签页不激活，不处理它

        if (document.visibilityState !== 'hidden') {
          // 设置预览界面
          setAddStatus('result');
          setSucceededBlock(msg.data);
        } else {
          setAddStatus('form');
        }
      }
    });
    /**
     * 失败之后清理状态
     * 应该保留日志，所以进行页面的切换
     */

    api.listenRemote({
      type: 'org.umi.block.add-blocks-fail',
      onMessage: function onMessage() {
        setTaskLoading(false);
        onAddBlockChange(undefined); // 如果标签页不激活，不处理它

        if (document.visibilityState !== 'hidden') {
          _antd.message.error(intl({
            id: 'org.umi.ui.blocks.adder.failed'
          }));
        }
      }
    });
    /**
     * 获取上次的安装的区块 url
     */

    callRemote({
      type: 'org.umi.block.get-adding-block-url'
    }).then(function (_ref2) {
      var data = _ref2.data;

      if (data) {
        // 如果有安装未完成的区块，设置显示页面为log
        // 并打开loading
        setAddStatus('log');
        setTaskLoading(true);
      }
    }); // detect language

    if (api.detectLanguage) {
      api.detectLanguage().then(function (language) {
        form.setFieldsValue({
          js: language === 'JavaScript'
        });
      });
    }
  }, []);
  (0, _react.useEffect)(function () {
    if (!block || !block.url) {
      return;
    } // 生成 defaultName


    var defaultName = block.url.split('/').pop();
    var initPath = blockType !== 'template' ? '/' : "/".concat(defaultName);

    var resetInitialValues = function resetInitialValues() {
      var noExitVar, initialValues;
      return regeneratorRuntime.async(function resetInitialValues$(_context2) {
        while (1) {
          switch (_context2.prev = _context2.next) {
            case 0:
              _context2.next = 2;
              return regeneratorRuntime.awrap((0, _util.getNoExitVar)({
                name: (0, _uppercamelcase.default)(defaultName),
                path: blockTarget || initPath,
                api: api,
                need: !!blockTarget
              }));

            case 2:
              noExitVar = _context2.sent;
              _context2.next = 5;
              return regeneratorRuntime.awrap((0, _util.getNoExitPath)({
                path: blockTarget || initPath,
                api: api,
                need: blockType === 'template'
              }));

            case 5:
              _context2.t0 = _context2.sent;
              _context2.next = 8;
              return regeneratorRuntime.awrap((0, _util.getNoExitRoute)({
                path: "/".concat(defaultName.toLocaleLowerCase()),
                api: api,
                need: blockType === 'template'
              }));

            case 8:
              _context2.t1 = _context2.sent;
              _context2.t2 = noExitVar;
              initialValues = {
                path: _context2.t0,
                routePath: _context2.t1,
                name: _context2.t2
              };
              form.setFieldsValue(initialValues);

            case 12:
            case "end":
              return _context2.stop();
          }
        }
      });
    };

    resetInitialValues();
  }, [block ? block.url : '', blockTarget || '']);
  (0, _react.useEffect)(function () {
    if (index !== null && index !== undefined) {
      form.setFieldsValue({
        index: index
      });
    }
  }, [index]);

  if (!block || !block.url) {
    return null;
  }
  /**
   * 计算初始值
   */


  var initialValues = {
    js: false,
    uni18n: localStorage.getItem('umi-ui-block-removeLocale') === 'true',
    npmClient: 'npm'
  };

  var renderOkText = function renderOkText(status, loading) {
    if (status === 'log' && !loading) {
      return intl({
        id: 'org.umi.ui.blocks.adder.stop'
      });
    }

    if (status === 'log') {
      return intl({
        id: 'org.umi.ui.blocks.adder.stop'
      });
    }

    return intl({
      id: 'org.umi.ui.blocks.adder.ok'
    });
  };

  return _react.default.createElement(_antd.Modal, {
    title: _react.default.createElement("div", {
      style: {
        display: 'flex'
      }
    }, intl({
      id: "org.umi.ui.blocks.adder.title.".concat(blockType)
    })),
    closable: true,
    visible: visible,
    destroyOnClose: true,
    onCancel: function onCancel() {
      onHideModal();

      if (!taskLoading) {
        setAddStatus('form');
      }
    },
    footer: addStatus === 'result' ? null : undefined,
    confirmLoading: fromCheck,
    bodyStyle: {
      height: '60vh',
      overflow: 'auto',
      transition: '.3s'
    },
    centered: true,
    okText: renderOkText(addStatus, taskLoading),
    onOk: function onOk() {
      if (addStatus === 'log' && !taskLoading) {
        onHideModal();
        setAddStatus('form');
        setSucceededBlock(undefined);
        return;
      }

      if (addStatus === 'log') {
        _antd.Modal.confirm({
          title: intl({
            id: 'org.umi.ui.blocks.adder.stop.title'
          }),
          content: intl({
            id: 'org.umi.ui.blocks.adder.stop.content'
          }),
          okType: 'danger',
          okText: intl({
            id: 'org.umi.ui.blocks.adder.stop.okText'
          }),
          cancelText: intl({
            id: 'org.umi.ui.blocks.adder.stop.cancelText'
          }),
          onOk: function onOk() {
            return regeneratorRuntime.async(function onOk$(_context3) {
              while (1) {
                switch (_context3.prev = _context3.next) {
                  case 0:
                    _context3.next = 2;
                    return regeneratorRuntime.awrap(cancelAddBlockTask(api));

                  case 2:
                    setTaskLoading(false);
                    onAddBlockChange(undefined);

                  case 4:
                  case "end":
                    return _context3.stop();
                }
              }
            });
          }
        });

        return;
      } // loading 状态更新


      setTaskLoading(true);
      setFromCheck(true);
      form.validateFields().then(function _callee(values) {
        var params;
        return regeneratorRuntime.async(function _callee$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                setAddStatus('log');
                _context4.t0 = _objectSpread;
                _context4.t1 = {};
                _context4.t2 = values;
                _context4.t3 = block.url;
                _context4.next = 7;
                return regeneratorRuntime.awrap((0, _BlockItem.getPathFromFilename)(api, values.path));

              case 7:
                _context4.t4 = _context4.sent;
                _context4.t5 = blockType === 'template' ? values.routePath : undefined;
                _context4.t6 = blockType === 'template';
                _context4.t7 = parseInt(values.index || '0', 0);
                _context4.t8 = blockType === 'template' ? block.name : values.name;
                _context4.t9 = {
                  url: _context4.t3,
                  path: _context4.t4,
                  routePath: _context4.t5,
                  page: _context4.t6,
                  index: _context4.t7,
                  name: _context4.t8
                };
                params = (0, _context4.t0)(_context4.t1, _context4.t2, _context4.t9);

                try {
                  addBlock(api, params);
                  localStorage.setItem('umi-ui-block-removeLocale', values.uni18n);
                  onAddBlockChange(block);
                } catch (error) {
                  _antd.message.error(error.message);
                }

              case 15:
              case "end":
                return _context4.stop();
            }
          }
        });
      }).catch(function () {
        return setTaskLoading(false);
      }).finally(function () {
        return setFromCheck(false);
      });
    }
  }, _react.default.createElement(_antd.Form, {
    hideRequiredMark: true,
    initialValues: initialValues,
    layout: "vertical",
    form: form,
    style: {
      display: addStatus !== 'form' && 'none'
    }
  }, blockType === 'template' && _react.default.createElement(_AddTemplateForm.default, {
    visible: visible,
    blockType: blockType
  }), blockType === 'block' && !api.isMini() && _react.default.createElement(_AddBlockForm.default, {
    form: form,
    visible: visible
  }), blockType === 'block' && api.isMini() && _react.default.createElement(_AddBlockFormForUI.default, {
    form: form,
    blockTarget: blockTarget
  }), _react.default.createElement(_antd.Form.Item, {
    name: "js",
    label: intl({
      id: 'org.umi.ui.blocks.adder.js'
    }),
    valuePropName: "checked",
    style: {
      display: 'none'
    }
  }, _react.default.createElement(_antd.Switch, {
    size: "small"
  })), blockType === 'template' && _react.default.createElement(_antd.Form.Item, {
    name: "uni18n",
    label: intl({
      id: 'org.umi.ui.blocks.adder.uni18n'
    }),
    valuePropName: "checked"
  }, _react.default.createElement(_antd.Switch, {
    size: "small"
  })), _react.default.createElement(_antd.Form.Item, {
    name: "npmClient",
    label: intl({
      id: 'org.umi.ui.blocks.adder.npmClient'
    })
  }, _react.default.createElement(_antd.Select, null, npmClients.map(function (client) {
    return _react.default.createElement(_antd.Select.Option, {
      key: client,
      value: client
    }, client);
  }))), _react.default.createElement(_antd.Form.Item, {
    name: "index"
  }, _react.default.createElement("input", {
    type: "hidden"
  }))), addStatus === 'log' && _react.default.createElement(_LogPanel.default, {
    loading: taskLoading
  }), addStatus === 'result' && succeededBlock && _react.default.createElement(_ResultPanel.default, {
    onFinish: function onFinish() {
      onHideModal();
      setAddStatus('form');
      setSucceededBlock(undefined);
      api.hideMini();
    },
    name: succeededBlock.name || block.name,
    url: succeededBlock.previewUrl
  }));
};

var _default = Adder;
exports.default = _default;
